pkg/expansion/system.go

```diff
@@ -0,0 +1,174 @@
+	b, err := json.MarshalIndent(v, "", "  ")
+	if err == nil {
+		return string(b) + "\n"
+	}
+	return ""
+}
+
+func genGVKToSchemaGVK(gvk expansionunversioned.GeneratedGVK) schema.GroupVersionKind {
+	return schema.GroupVersionKind{
+		Group:   gvk.Group,
+		Version: gvk.Version,
+		Kind:    gvk.Kind,
+	}
+}
+
+// templatesForGVK returns a slice of ExpansionTemplates that match a given gvk.
+func (s *System) templatesForGVK(gvk schema.GroupVersionKind) []*expansionunversioned.ExpansionTemplate {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	var templates []*expansionunversioned.ExpansionTemplate
+	for _, t := range s.templates {
+		for _, apply := range t.Spec.ApplyTo {
+			if apply.Matches(gvk) {
+				templates = append(templates, t)
+			}
+		}
+	}
+
+	return templates
+}
+
+// Expand expands `base` into resultant resources, and applies any applicable
+// mutators. If no ExpansionTemplates match `base`, an empty slice
+// will be returned. If `s.mutationSystem` is nil, no mutations will be applied.
+func (s *System) Expand(base *mutationtypes.Mutable) ([]*unstructured.Unstructured, error) {
+	gvk := base.Object.GroupVersionKind()
+	if gvk == (schema.GroupVersionKind{}) {
+		return nil, fmt.Errorf("cannot expand resource %s with empty GVK", base.Object.GetName())
+	}
+	var resultants []*unstructured.Unstructured
+	templates := s.templatesForGVK(gvk)
+
+	for _, te := range templates {
+		res, err := expandResource(base.Object, base.Namespace, te)
+		resultants = append(resultants, res)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if s.mutationSystem == nil {
+		return resultants, nil
+	}
+
+	for _, res := range resultants {
+		mutable := &mutationtypes.Mutable{
+			Object:    res,
+			Namespace: base.Namespace,
+			Username:  base.Username,
+			Source:    mutationtypes.SourceTypeGenerated,
+		}
+		_, err := s.mutationSystem.Mutate(mutable)
+		if err != nil {
+			return nil, fmt.Errorf("failed to mutate resultant resource %s: %s", res.GetName(), err)
+		}
+	}
+
+	return resultants, nil
+}
+
+func expandResource(obj *unstructured.Unstructured, ns *corev1.Namespace, template *expansionunversioned.ExpansionTemplate) (*unstructured.Unstructured, error) {
+	srcPath := template.Spec.TemplateSource
+	if srcPath == "" {
+		return nil, fmt.Errorf("cannot expand resource using a template with no source")
+	}
+	resultantGVK := genGVKToSchemaGVK(template.Spec.GeneratedGVK)
+	if resultantGVK == (schema.GroupVersionKind{}) {
+		return nil, fmt.Errorf("cannot expand resource using template with empty generatedGVK")
+	}
+
+	src, ok, err := unstructured.NestedMap(obj.Object, sourcePath(srcPath)...)
+	if err != nil {
+		return nil, fmt.Errorf("could not extract source field from unstructured: %s", err)
+	}
+	if !ok {
+		return nil, fmt.Errorf("could not find source field %q in obj", srcPath)
+	}
+
+	resource := &unstructured.Unstructured{}
+	resource.SetUnstructuredContent(src)
+	resource.SetGroupVersionKind(resultantGVK)
+	resource.SetNamespace(ns.Namespace)
```

Comment:
ns.Name? I'm not sure Namespace is set on namespaces, particularly if this is read from disk.